ddtrace/tracer/textmap.go

```diff
@@ -462,3 +475,137 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
 	}
 	return &ctx, nil
 }
+
+const (
+	traceparentHeader = "traceparent"
+	tracestateHeader  = "tracestate"
+)
+
+// propagatorW3c implements Propagator and injects/extracts span contexts
+// using W3C tracecontext/traceparent headers. Only TextMap carriers are supported.
+type propagatorW3c struct{}
+
+func (p *propagatorW3c) Inject(context ddtrace.SpanContext, carrier interface{}) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *propagatorW3c) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
+	switch c := carrier.(type) {
+	case TextMapReader:
+		return p.extractTextMap(c)
+	default:
+		return nil, ErrInvalidCarrier
+	}
+}
+
+func (*propagatorW3c) extractTextMap(reader TextMapReader) (ddtrace.SpanContext, error) {
+	var parentHeader string
+	stateHeaders := []string{}
+	// to avoid parsing tracestate header(s) if traceparent is invalid
+	err := reader.ForeachKey(func(k, v string) error {
+		key := strings.ToLower(k)
+		switch key {
+		case traceparentHeader:
+			if parentHeader != "" {
+				return ErrSpanContextCorrupted
+			}
+			parentHeader = v
+		case tracestateHeader:
+			stateHeaders = append(stateHeaders, v)
+		}
+		return nil
+	})
+	if err != nil || parentHeader == "" {
```

Comment:
Is it valid if there are no tracestate headers?

Reply:
We're also checking the length of `parentHeader` in `parseTraceparent` already (after we trim it), so removing this check for `parentHeader == ""` here would still work fine and return the error we expect

Reply:
Comment N1: From RFC: 
If the traceparent is valid, parse the tracestate header with the following logic. If the tracestate header is not present, use the available traceparent information to create a span context.
CommentN2: 👍